examples/hmm_enum.py

```diff
@@ -0,0 +1,279 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Enumerate Hidden Markov Model
+=============================
+
+This example is ported from [1], which shows how to marginalize out
+discrete model variables in Pyro.
+
+This combines MCMC with a variable elimination algorithm, where we
+use enumeration to exactly marginalize out some variables from the
+joint density.
+
+To marginalize out discrete variables ``x``:
+
+1. Verify that the variable dependency structure in your model
+    admits tractable inference, i.e. the dependency graph among
+    enumerated variables should have narrow treewidth.
+2. Ensure your model can handle broadcasting of the sample values
+    of those variables
+
+Note that difference from [1], which uses Python loop, here we use
+:func:`~numpryo.contrib.control_flow.scan` to reduce compilation
+times (only one step needs to be compiled) of the model. Under the
+hood, `scan` stacks all the priors' parameters and values into
+an additional time dimension. This allows us computing the joint
+density in parallel. In addition, the stacked form allows us
+to use the parallel-scan algorithm in [2], which reduces parallel
+complexity from O(length) to O(log(length)).
+
```

Comment:
@eb8680 Could you suggest me a paragraph mentioning about how to obtain the data and how to preprocess it?

Reply:
I think the data you're using is a single batch from the music dataset used in the Pyro HMM and DMM examples. The relevant Pyro file, with a bit of commentary, is [here](https://github.com/pyro-ppl/pyro/blob/dev/examples/dmm/polyphonic_data_loader.py).  Unfortunately the preprocessing logic there depends heavily on PyTorch, so the easiest thing to do is probably to do the preprocessing once in Pyro, save the resulting transformed data, and store it in Cloudfront.

Reply:
I see. I think we just need to mention what is the structure of the data. E.g. in `pyro`, the data has 88 tones but it seems to only have 44 tones here.

Reply:
I think there are only 44 tones here because the batch I grabbed only happened to contain 44 notes.  I would suggest using the full dataset instead now that the compilation issues have been resolved.

Reply:
I can confirm that there are 44 notes with `num_sequences=17`. :D